File: backend/app/models/product.py
"""Product-related Pydantic schemas for inventory endpoints."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class ProductCreate(BaseModel):
    """Payload for creating a unified inventory product."""

    name: str
    brand: str
    category: str
    description: str
    price: float
    totalStock: int = Field(ge=0)
    platforms: List[str]
    isManuallyReviewed: bool = True
    matchingConfidence: str = "high"
    sku: Optional[str] = None
    primaryImageUrl: Optional[str] = None


class ProductUpdate(BaseModel):
    """Partial product update payload."""

    name: Optional[str] = None
    brand: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    totalStock: Optional[int] = Field(default=None, ge=0)
    platforms: Optional[List[str]] = None
    isManuallyReviewed: Optional[bool] = None
    matchingConfidence: Optional[str] = None
    sku: Optional[str] = None
    primaryImageUrl: Optional[str] = None


class ProductResponse(ProductCreate):
    """Unified product returned to API consumers."""

    id: str
    createdAt: datetime
    updatedAt: datetime


class StockAdjustmentRequest(BaseModel):
    """Manual stock adjustment with audit metadata."""

    productId: str
    productName: str
    adjustmentType: str  # "increase" | "decrease"
    quantity: int = Field(gt=0)
    oldStock: int
    newStock: int = Field(ge=0)
    reason: str
    notes: str = ""
    adjustmentDate: datetime

File: lib/data/models/product.dart
/// Unified catalog product model deserialized from FastAPI inventory endpoints.
class Product {
  const Product({
    required this.id,
    required this.name,
    required this.brand,
    required this.category,
    required this.description,
    required this.price,
    required this.totalStock,
    required this.platforms,
    this.sku,
    this.primaryImageUrl,
    this.isManuallyReviewed = true,
    this.matchingConfidence = 'high',
  });

  final String id;
  final String name;
  final String brand;
  final String category;
  final String description;
  final double price;
  final int totalStock;
  final List<String> platforms;
  final String? sku;
  final String? primaryImageUrl;
  final bool isManuallyReviewed;
  final String matchingConfidence;

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      name: json['name'] as String? ?? '',
      brand: json['brand'] as String? ?? '',
      category: json['category'] as String? ?? '',
      description: json['description'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0,
      totalStock: (json['totalStock'] as num?)?.toInt() ?? 0,
      platforms: (json['platforms'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      sku: json['sku'] as String?,
      primaryImageUrl: json['primaryImageUrl'] as String?,
      isManuallyReviewed: json['isManuallyReviewed'] as bool? ?? true,
      matchingConfidence: json['matchingConfidence'] as String? ?? 'high',
    );
  }

  String get displayLabel {
    if (brand.isNotEmpty) {
      return '$name ($brand) - Stock: $totalStock';
    }
    return '$name - Stock: $totalStock';
  }
}
